basex-core/src/main/java/org/basex/core/locks/DBLocking.java

```diff
@@ -138,7 +138,9 @@ public void acquire(final Proc pr, final StringList read, final StringList write
       }
       // global read locking
       if(read == null) {
-        while(localWriters > 0) {
+        while(localWriters > 0 &&
+            // We're the only writer, allow global read lock anyway
+            !(1 == localWriters && !(null == write || write.isEmpty()))) {
```

Comment:
I'm trying to understand what's going on in here.. So the loop is either interrupted if:

1. `localWriter = 0`, or
2. `localWriter = 1` and:
  1. the write lock is global, or
  2. there are no local write locks

Did I get it right? If yes, the code could possible be simplified a bit as follows:

```java
while(localWriters > 0 && (localWriters > 2 || write == null || write.isEmpty()))
```